stl/inc/iterator

```diff
@@ -513,6 +513,501 @@ private:
+            case _Variantish_state::_Holds_sentinel:
+                _Sentinel = _STD move(_That._Sentinel);
+                break;
+            case _Variantish_state::_Nothing:
+                break;
+            }
+
+            return *this;
+        }
+
+        _Clear();
+        _Contains = _Variantish_state::_Nothing;
+
+        switch (_That._Contains) {
+        case _Variantish_state::_Holds_iter:
+            _Construct_in_place(_Iterator, _STD move(_That._Iterator));
+            break;
+        case _Variantish_state::_Holds_sentinel:
+            _Construct_in_place(_Sentinel, _STD move(_That._Sentinel));
+            break;
+        case _Variantish_state::_Nothing:
+            break;
+        }
+        _Contains = _That._Contains;
+
+        return *this;
+    }
+
+    // clang-format off
+    template <class _OIter, class _OSe>
+        requires _Not_same_as<_Variantish<_OIter, _OSe>, _Variantish>
+    _Variantish& operator=(const _Variantish<_OIter, _OSe>& _That) noexcept(
+        is_nothrow_convertible_v<const _OIter&, _It>&& is_nothrow_convertible_v<const _OSe&, _Se>) {
+        // clang-format on
+        if (_Contains == _That._Contains) {
+            switch (_Contains) {
+            case _Variantish_state::_Holds_iter:
+                _Iterator = _That._Iterator;
+                break;
+            case _Variantish_state::_Holds_sentinel:
+                _Sentinel = _That._Sentinel;
+                break;
+            case _Variantish_state::_Nothing:
+                break;
+            }
+
+            return *this;
+        }
+
+        _Clear();
+        _Contains = _Variantish_state::_Nothing;
+
+        switch (_That._Contains) {
+        case _Variantish_state::_Holds_iter:
+            _Construct_in_place(_Iterator, _That._Iterator);
+            break;
+        case _Variantish_state::_Holds_sentinel:
+            _Construct_in_place(_Sentinel, _That._Sentinel);
+            break;
+        case _Variantish_state::_Nothing:
+            break;
+        }
+
+        _Contains = _That._Contains;
+
+        return *this;
+    }
+
+    // clang-format off
+    constexpr friend void swap(_Variantish& _Left, _Variantish& _Right)
+        requires (!_Is_trivially_swappable_v<_It> || !_Is_trivially_swappable_v<_Se>) {
+        // clang-format on
+        if (_Left._Contains == _Right._Contains) {
+            switch (_Left._Contains) {
+            case _Variantish_state::_Holds_iter:
+                _RANGES swap(_Left._Iterator, _Right._Iterator);
+                break;
+            case _Variantish_state::_Holds_sentinel:
+                _RANGES swap(_Left._Sentinel, _Right._Sentinel);
+                break;
+            case _Variantish_state::_Nothing:
+                break;
+            }
+
+            return;
+        }
+
+        auto _Tmp = _STD move(_Left);
+        _Left     = _STD move(_Right);
+        _Right    = _STD move(_Tmp);
+    }
+
+    void _Clear() noexcept {

```

Comment:
I've never seen MSVC's optimizer eliminate a store in a destructor for an escaped object (https://godbolt.org/z/79WP87). Compromise position: I'll rename this `_Raw_clear`, and add a `_Clear` that calls it and also sets `_Contains`. This will nicely simply the callers.